### PR TITLE
[SemaCXX] Implement CWG2137 (list-initialization from objects of the same type)

### LINKED ISSUE #77638: [D156032] Implement CWG2137 (list-initialization from objects of the same type)
This issue tracks picking up https://reviews.llvm.org/D156032 from the Phabricator archive, or dropping it in case it has already landed. I want to make sure nothing gets lost.

### PR DESCRIPTION
**Pull request by MitalAshok**
Closes #77638, #24186

Rebased from <https://reviews.llvm.org/D156032>, see there for more information.

Implements wording change in [CWG2137](https://wg21.link/CWG2137) in the first commit.

This also implements an approach to [CWG2311](https://wg21.link/CWG2311) in the second commit, because too much code that relies on `T{ T_prvalue }` being an elision would break. Because that issue is still open and the CWG issue doesn't provide wording to fix the issue, there may be different behaviours on other compilers.

### CODE REVIEW

**Review by cor3ntin — APPROVED** [review]
I think all the comments made on phab have been applied, and I do believe this implements the DRs (or their intent)

Please give a few days for other folks to look at it before merging.
@zygoloid in particular as he opened CW2311

**Comment by cor3ntin** [comment]
@mordante ping for visibility (you approved that on Phab already)

File: clang/test/CXX/drs/dr21xx.cpp
Comment (Endilll) [inline clang/test/CXX/drs/dr21xx.cpp:153]: Can you redo `expected` directives, so that they are consistent with the rest of the file?

**Review by Endilll — CHANGES_REQUESTED** [review]

**Review by mordante — APPROVED** [review]
Thanks @cor3ntin. The patch still looks fine from libc++'s PoV.

File: clang/test/CXX/drs/dr21xx.cpp
Comment (Endilll) [inline clang/test/CXX/drs/dr21xx.cpp:156]: Can you follow the example of test for CWG2141? Specifically,
1. not trimming diagnostic messages (including identifiers);
2. placing expected errors on the following line (via `// since-cxx11-error@-1`);
3. grouping notes together with errors the same way they are grouped in console output (2170 is a good example how to handle notes that point to a different source line).

**Review by Endilll — CHANGES_REQUESTED** [review]

File: clang/test/CXX/drs/dr21xx.cpp
Comment (MitalAshok) [reply clang/test/CXX/drs/dr21xx.cpp]: @Endilll Is this good now? Thanks in advance

https://github.com/llvm/llvm-project/pull/77768/files#diff-3426b975672f064b9f32caf4ebab729b815f4f65aa3e76bda4b5bdee986e2e6fR130-R131

https://github.com/llvm/llvm-project/pull/77768/files#diff-8807231ab4d193692d6cecdc6428aaf3c55dfd96101cff8fe0eb7034654658fbR154-R161

**Review by Endilll — APPROVED** [review]
Looks perfect to me now. Thank you!

**Comment by cor3ntin** [comment]
@MitalAshok the bot found some trailing whitespace issues

**Comment by MitalAshok** [comment]
@cor3ntin It looks like it was in an unrelated file. I've rebased and the format check is passing now

**Comment by cor3ntin** [comment]
Do you want me to merge on your behalf?

**Comment by MitalAshok** [comment]
@cor3ntin Still waiting for Windows checks, but yes, could you merge if you please

**Comment by cor3ntin** [comment]
Let me merge that lest I forget. Our windows bots have a lot of issues

**Comment by alexfh** [comment]
We've started seeing `call to constructor of 'SomeType' is ambiguous` errors after this patch on some code. In the case I closely inspected the code looks like https://gcc.godbolt.org/z/3MbWf69cx:

```
#include <vector>
struct A {
    A() {}
};

struct B {
    explicit B(A);
    explicit B(std::vector<A>);
};

struct C {
    B b;
    C() : b({A()}) {}
};
```

Is the new error expected? (I would totally agree with it, just making sure. And the behavior is now different to that of GCC trunk.)

**Comment by alexfh** [comment]
Actually, on the second look it doesn't seem to be too ambiguous to me. I *think* `std::vector<A>` should be a strictly better match in this case.

**Comment by cor3ntin** [comment]
@MitalAshok Could you look at that today? Otherwise we need to revert - both in main and in the 18 branch

**Comment by alexfh** [comment]
Given the broad impact we see in our code, I suppose this can affect other users as well. I suggest to revert first and then find a proper solution.

**Comment by MitalAshok** [comment]
With the standard as is, `B(A)` should be a better match because of [[over.ics.list]p(7.1)](https://wg21.link/over.ics.list#7.1), it is an Exact Match, and `B(std::vector<A>)` is a user defined conversion.

With the current CWG2311 fix, the call to `B(A)` is an Exact Match because it's an `A` prvalue.

https://github.com/llvm/llvm-project/blob/7b11c08c664863fbcd7a3058179b0af3de5d28e4/clang/lib/Sema/SemaOverload.cpp#L1593

^ This line is the issue: `isCopyConstructor()` checks only for lvalue reference copy constructors, and the move constructor is selected (i.e., if you force a copy constructor to be called instead, it works: https://godbolt.org/z/rGEsPdMx8)

Removing `Constructor->isCopyConstructor() &&` fixes the issue (`B(A)` is chosen, and all existing tests still pass)

@cor3ntin I have some extra test cases I would want to add too. Should I make a new pull request? Or are we going to reopen this one and target llvm 19?

**Comment by cor3ntin** [comment]
@MitalAshok Please make a new PR!
We can then cherry-pick it to 18

**Comment by alexfh** [comment]
I'll go ahead and revert the commit for now. My local test run has just finished with no new errors.

**Comment by erichkeane** [comment]
> I'll go ahead and revert the commit for now. My local test run has just finished with no new errors. Feel free to recommit with the new test cases and the fix.

Doing a revert like this during our release process is actually quite a disturbance, particularly when the committer is actively working on the fix (and has identified it!).  In the future please give them a chance to fix it before reverting on them.

ALSO, in the future, if you are going to do a revert while we are in a release process, you need to make sure to submit a request to revert it from the release branch at the same time, you cannot just revert in 'main'.

**Comment by alexfh** [comment]
> Doing a revert like this during our release process is actually quite a disturbance, 

My apologies, I didn't realize the additional complexities due to the release.

> particularly when the committer is actively working on the fix (and has identified it!). In the future please give them a chance to fix it before reverting on them.

As I understood the author, more time was needed to prepare the tests and the fix. I just tried to bring the tip of the tree to a good state soon (our internal release process was blocked on this) without putting any pressure on the author.

> ALSO, in the future, if you are going to do a revert while we are in a release process, you need to make sure to submit a request to revert it from the release branch at the same time, you cannot just revert in 'main'.

Should I send a PR to revert the commit in the release branch now?

**Comment by erichkeane** [comment]
>As I understood the author, more time was needed to prepare the tests and the fix. I just tried to bring the tip of the tree to a good state soon (our internal release process was blocked on this) without putting any pressure on the author.

Typically when the author is engaging fairly quickly and only needs a short time (he had already identified the problem and only had process-related concerns), it is preferred to be understanding/polite and give them time to work on it.

While I am sympathetic to your internal processes, it seems to me that they should be more tolerant of a ToT that is not particularly stable, as is the case with Clang.

> With the standard as is, `B(A)` should be a better match because of [[over.ics.list]p(7.1)](https://wg21.link/over.ics.list#7.1), it is an Exact Match, and `B(std::vector<A>)` is a user defined conversion.
> 
> With the current CWG2311 fix, the call to `B(A)` is an Exact Match because it's an `A` prvalue.
> 
> In either case, `B(A)` should unambiguously be chosen over `B(std::vector<A>)` given `{ A() }` as an initializer.
> 
> https://github.com/llvm/llvm-project/blob/7b11c08c664863fbcd7a3058179b0af3de5d28e4/clang/lib/Sema/SemaOverload.cpp#L1593
> 
> ^ This line is the issue: `isCopyConstructor()` checks only for lvalue reference copy constructors, and the move constructor is selected (i.e., if you force a copy constructor to be called instead, it works: https://godbolt.org/z/rGEsPdMx8).
> 
> There is no reason to check for copy constructors here. Removing `Constructor->isCopyConstructor() &&` fixes the issue (`B(A)` is chosen, and all existing tests still pass)
> 
> @cor3ntin I have some extra test cases I would want to add too. Should I make a new pull request? Or are we going to reopen this one and target llvm 19?

If you haven't already, you can re-open this one.  However, you'll ALSO have to make a 'differential' PR against 18.x so that this can be fixed there too.

**Comment by alexfh** [comment]
> > As I understood the author, more time was needed to prepare the tests and the fix. I just tried to bring the tip of the tree to a good state soon (our internal release process was blocked on this) without putting any pressure on the author.
> 
> Typically when the author is engaging fairly quickly and only needs a short time (he had already identified the problem and only had process-related concerns), it is preferred to be understanding/polite and give them time to work on it.

First of all, let me quote the developer policy: ["Remember, it is normal and healthy to have patches reverted. Having a patch reverted does not necessarily mean you did anything wrong."](https://llvm.org/docs/DeveloperPolicy.html#patch-reversion-policy). Now, I totally understand the desire to "capture the progress" as well as frustration one might have seeing their work reverted. However, a revert shouldn't be perceived as an impolite or unfriendly action. All contributions are welcome and appreciated. Reverting a commit is merely a way to deal with problems one at a time without blocking others or racing against the clock.

Frequently, it is the most sustainable way forward that helps keeping the tip of the tree in a reasonably good shape while not putting unnecessary time pressure on contributors. ["As a community, we strongly value having the tip of tree in a good state while allowing rapid iterative development. As such, we tend to make much heavier use of reverts to keep the tree healthy than some other open source projects, and our norms are a bit different."](https://llvm.org/docs/DeveloperPolicy.html#patch-reversion-policy)

A quick and obvious forward fix is also an option sometimes, but though there are certain signals like fast responses from all involved parties, the time to get a **correct** fix into the tree is unpredictable in general. Sometimes, an "obvious" fix turns out to be suboptimal or just wrong, which can be found during a review or much later - after it's committed and tested (by us or other users). For example, I got a quick response here, but now, two days later the fix hasn't been even sent for review (and this is totally fine, since the problematic commit is reverted, and the author can work in their pace and with their own priorities).

Things get even more complicated, when a problem is long enough in the tree and other problems start to pile up. It's a matter of just two-three sufficiently widespread issues being simultaneously in the tree, when even finding the culprit for each problem becomes much longer and requires more effort. Been there, done that. For example, this week we've already seen a handful of clang bugs affecting the same files, leading to them be hidden behind each other.

> While I am sympathetic to your internal processes, it seems to me that they should be more tolerant of a ToT that is not particularly stable, as is the case with Clang.

You may be surprised, but ToT Clang (as well as LLVM in general) is quite stable. Our release process ensures that every week or so there is an LLVM revision that requires just a handful of cherrypicks (sometimes, no cherrypicks at all) to correctly compile all of our code, which, given the immense size, is a pretty high bar (though, this obviously applies only to the configurations and targets we use).

**Comment by AaronBallman** [comment]
> > > As I understood the author, more time was needed to prepare the tests and the fix. I just tried to bring the tip of the tree to a good state soon (our internal release process was blocked on this) without putting any pressure on the author.
> > 
> > 
> > Typically when the author is engaging fairly quickly and only needs a short time (he had already identified the problem and only had process-related concerns), it is preferred to be understanding/polite and give them time to work on it.
> 
> First of all, let me quote the developer policy: ["Remember, it is normal and healthy to have patches reverted. Having a patch reverted does not necessarily mean you did anything wrong."](https://llvm.org/docs/DeveloperPolicy.html#patch-reversion-policy). Now, I totally understand the desire to "capture the progress" as well as frustration one might have seeing their work reverted. However, a revert shouldn't be perceived as an impolite or unfriendly action. All contributions are welcome and appreciated. Reverting a commit is merely a way to deal with problems one at a time without blocking others or racing against the clock.

The developer policy also says: "Reverts should be reasonably timely. A change submitted two hours ago can be reverted without prior discussion. A change submitted two years ago should not be. Where exactly the transition point is is hard to say, but it’s probably in the handful of days in tree territory. If you are unsure, we encourage you to reply to the commit thread, give the author a bit to respond, and then proceed with the revert if the author doesn’t seem to be actively responding."

"Reasonably timely" also means that authors who are engaged on a fix are expected to have the opportunity to land that fix. Reverts can be disruptive and they can clutter the revision history, so *if* we can avoid them, it's better to avoid them. However, we don't want to avoid reverts when they help keep the tree stable; not reverting can be plenty disruptive, too! There's a happy medium somewhere between "revert as quickly as possible at the first sign of trouble" and "let authors have infinite time to fix issues".

I think it's better that we err on the side of reverting quickly, and I don't think anyone has done anything wrong here. But I think we have some downstream projects which are quicker to revert than is comfortable and so some code owners are pushing back (gently) on reverts that we think may have jumped the gun a bit. For example, I know I've personally had several instances where I've had things reverted out from under me as I'm trying to land the fix and that's caused significantly more work for me. This review seems to have had a similar outcome: the community was told about the issue, the patch author immediately responded, and the patch revert happened regardless and more effort is required from everyone as a result. Again, I still think the revert is defensible. But I also think the push back on the speed at which the revert happened is defensible as well. This is mostly just a call to coordinate a bit better. I would suggest that if the patch author is engaged on the thread, it would make sense to tell them the schedule on which you plan to revert so they have an opportunity to tell you if that's going to be disruptive.

**Comment by DimitryAndric** [comment]
FWIW I think I just hit this when building llvm itself (though 15.0.6) with a recent clang, and getting:

```text
/wrkdirs/usr/ports/devel/llvm15/work-default/llvm-project-15.0.7.src/llvm/tools/sancov/sancov.cpp:532:44: error: chosen constructor is explicit in copy-initialization
  532 |       return SpecialCaseList::createOrDie({{ClBlacklist}},
      |                                            ^~~~~~~~~~~~~
/usr/include/c++/v1/string:1047:93: note: explicit constructor declared here
 1047 |   _LIBCPP_METHOD_TEMPLATE_IMPLICIT_INSTANTIATION_VIS _LIBCPP_CONSTEXPR_SINCE_CXX20 explicit basic_string(const _Tp& __t)
      |                                                                                             ^
/usr/include/c++/v1/vector:509:91: note: passing argument to parameter '__il' here
  509 |   _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI vector(initializer_list<value_type> __il);
      |                                                                                           ^
/wrkdirs/usr/ports/devel/llvm15/work-default/llvm-project-15.0.7.src/llvm/tools/sancov/sancov.cpp:535:42: error: chosen constructor is explicit in copy-initialization
  535 |     return SpecialCaseList::createOrDie({{ClIgnorelist}},
      |                                          ^~~~~~~~~~~~~~
/usr/include/c++/v1/string:1047:93: note: explicit constructor declared here
 1047 |   _LIBCPP_METHOD_TEMPLATE_IMPLICIT_INSTANTIATION_VIS _LIBCPP_CONSTEXPR_SINCE_CXX20 explicit basic_string(const _Tp& __t)
      |                                                                                             ^
/usr/include/c++/v1/vector:509:91: note: passing argument to parameter '__il' here
  509 |   _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI vector(initializer_list<value_type> __il);
      |                                                                                           ^
2 errors generated.
```

I didn't managed to fully bisect it yet, but it looks very much like 924701311aa79180e86ad8ce43d253f27d25ec7d for this issue is the culprit.

So has that now been reverted, and is the reversal also merged into the 18.x release branch?

**Comment by erichkeane** [comment]
> However, a revert shouldn't be perceived as an impolite or unfriendly action. All contributions are welcome and appreciated. Reverting a commit is merely a way to deal with problems one at a time without blocking others or racing against the clock.

Unfortunately this isn't the way it is perceived.  No matter how much we say that, it causes folks to get discouraged/scares them away.  I've witnessed a significant number of 'new' contributors do great for 3-4 patches, then have their 'last' patch get reverted, and that person disappears from the project.  I know when _I_ first got reverted (a long time ago now!) I nearly quit the project entirely, and I had many coworkers tell me they no longer wanted to contribute to clang after getting reverted.

While I realize this is the 'policy', I'm asking for compassion, particularly with new developers who are engaging.

>For example, I got a quick response here, but now, two days later the fix hasn't been even sent for review (and this is totally fine, since the problematic commit is reverted, and the author can work in their pace and with their own priorities).

I find myself wondering if/afraid that the 'revert' is why we haven't seen them come back, and not because they now have 'more time'.

While I am sympathetic to the goals of keeping ToT green, I just asked that you be equally sympathetic to the person you're reverting.

**Comment by alexfh** [comment]
> This is mostly just a call to coordinate a bit better. I would suggest that if the patch author is engaged on the thread, it would make sense to tell them the schedule on which you plan to revert so they have an opportunity to tell you if that's going to be disruptive.

This part captures one particular issue with my way of action I happened to miss. Thanks for putting this clearly @AaronBallman! Coordination was indeed far from perfect here. And your suggestion totally makes sense.

> While I realize this is the 'policy', I'm asking for compassion, particularly with new developers who are engaging.

This is also a good point @erichkeane. I'll try to consider this as well.

**Comment by yxsamliu** [comment]
This PR caused a regression for AMDMIGraphX https://github.com/ROCm/AMDMIGraphX/blob/develop/src/targets/gpu/include/migraphx/gpu/prefix_scan_sum.hpp#L57

A reduced test case is:

#include <vector>
#include <stdio.h>
using namespace std;
```
struct A {
    int x;
    A(int x_) : x(x_) {}
    A(const vector<A>& a) {}
};

int main() {
    vector<A> a{1,2};
    vector<A> b{a};

    printf("%ld\n", b.size());
}
```
For variable b, it is expected to be constructed by copy ctor of vector, and have 2 elements.

With this patch, a is first converted to A(a), then b is constructed with one element A(a).

Is this expected or a bug? Thanks.

gcc has the same behavior https://godbolt.org/z/hTb795xWP

**Comment by yxsamliu** [comment]
> > This is mostly just a call to coordinate a bit better. I would suggest that if the patch author is engaged on the thread, it would make sense to tell them the schedule on which you plan to revert so they have an opportunity to tell you if that's going to be disruptive.
> 
> This part captures one particular issue with my way of action I happened to miss. Thanks for putting this clearly @AaronBallman! Coordination was indeed far from perfect here. And your suggestion totally makes sense.
> 
> > While I realize this is the 'policy', I'm asking for compassion, particularly with new developers who are engaging.
> 
> This is also a good point @erichkeane. I'll try to consider this as well.

Any plan to reland this PR? cuDF/hipDF depends on it. Also the subtle difference of behavior about list initialization between clang/gcc causes issues that are difficult to debug. Thanks.

**Pull request merged**
